Sources/HummingbirdWSCore/PerMessageDeflateExtension.swift

```diff
@@ -0,0 +1,278 @@
+        let compressionLevel: Int?
+        let memoryLevel: Int?
+    }
+
+    /// Internal mutable state and referenced types, that cannot be set to Sendable
+    class InternalState {
+        fileprivate let decompressor: any NIODecompressor
+        fileprivate let compressor: any NIOCompressor
+        fileprivate var sendState: SendState
+
+        init(configuration: Configuration) throws {
+            self.decompressor = CompressionAlgorithm.deflate(
+                configuration: .init(
+                    windowSize: numericCast(configuration.receiveMaxWindow ?? 15)
+                )
+            ).decompressor
+            // compression level -1 will setup the default compression level, 8 is the default memory level
+            self.compressor = CompressionAlgorithm.deflate(
+                configuration: .init(
+                    windowSize: numericCast(configuration.sendMaxWindow ?? 15),
+                    compressionLevel: configuration.compressionLevel.map { numericCast($0) } ?? -1,
+                    memoryLevel: configuration.memoryLevel.map { numericCast($0) } ?? 8
+                )
+            ).compressor
+            self.sendState = .idle
+            try self.decompressor.startStream()
+            try self.compressor.startStream()
+        }
+
+        func shutdown() {
+            try? self.compressor.finishStream()
+            try? self.decompressor.finishStream()
+        }
+    }
+
+    let configuration: Configuration
+    let internalState: NIOLoopBound<InternalState>
+
+    init(configuration: Configuration, eventLoop: EventLoop) throws {
+        self.configuration = configuration
+        self.internalState = try .init(.init(configuration: configuration), eventLoop: eventLoop)
+    }
+
+    func shutdown() {
+        self.internalState.value.shutdown()
+    }
+
+    func processReceivedFrame(_ frame: WebSocketFrame, ws: HBWebSocket) throws -> WebSocketFrame {
+        var frame = frame
+        if frame.rsv1 {
+            let state = self.internalState.value
+            precondition(frame.fin, "Only concatenated frames with fin set can be processed by the permessage-deflate extension")
+            frame.data.writeBytes([0, 0, 255, 255])
+            frame.data = try frame.data.decompressStream(with: state.decompressor, maxSize: ws.maxFrameSize, allocator: ws.channel.allocator)
+            if self.configuration.receiveNoContextTakeover {
+                try state.decompressor.resetStream()
+            }
+        }
+        return frame
+    }
+
+    func processFrameToSend(_ frame: WebSocketFrame, ws: HBWebSocket) throws -> WebSocketFrame {
+        // if the frame is larger than 16 bytes, we haven't received a final frame or we are in the process of sending a message
+        // compress the data
+        let state = self.internalState.value
+        if frame.data.readableBytes > 16 || !frame.fin || state.sendState != .idle, frame.opcode == .text || frame.opcode == .binary {
```

Comment:
Can we break this if-statement into two expressions as variables, and **then** combine them?